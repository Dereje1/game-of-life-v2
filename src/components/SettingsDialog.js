import React from "react";
import PropTypes from "prop-types";
import DialogContent from "@mui/material/DialogContent";
import DialogActions from "@mui/material/DialogActions";
import Dialog from "@mui/material/Dialog";
import IconButton from "@mui/material/IconButton";
import CancelOutlinedIcon from "@mui/icons-material/CancelOutlined";
import ListItem from "@mui/material/ListItem";
import ListItemText from "@mui/material/ListItemText";
import ListItemIcon from "@mui/material/ListItemIcon";
import HelpIcon from "@mui/icons-material/Help";
import GitHubIcon from "@mui/icons-material/GitHub";
import Switch from "@mui/material/Switch";
import Grid4x4Icon from "@mui/icons-material/Grid4x4";
import Typography from "@mui/material/Typography";
import InfoIcon from "@mui/icons-material/Info";
import PatternIcon from "@mui/icons-material/Pattern";
import RefreshIcon from "@mui/icons-material/Refresh";
import FormatColorFillIcon from "@mui/icons-material/FormatColorFill";
import ColorLensIcon from "@mui/icons-material/ColorLens";
import patterns from "../utils/patterns";

const getPatternLabel = (patternName) => {
  const [pattern] = patterns.filter((p) => p.value === patternName);
  return pattern.label;
};

const SettingsDialog = ({
  open,
  values,
  handleOk,
  showGrid,
  handleGrid,
  patternName,
  refreshPattern,
  handleColorPicker
}) => (
  <Dialog
    sx={{ "& .MuiDialog-paper": { width: "80%", maxHeight: 530 } }}
    maxWidth="xs"
    open={open}
    onClose={handleOk}
  >
    <DialogContent dividers>
      <ListItem>
        <ListItemIcon>
          <PatternIcon color={patternName === "none" ? "" : "info"} />
        </ListItemIcon>
        <ListItemText id="switch-list-label-grid" primary={getPatternLabel(patternName)} />
        <IconButton disabled={patternName === "none"} onClick={refreshPattern}>
          <RefreshIcon color={patternName === "none" ? "" : "success"} />
        </IconButton>
      </ListItem>
      <ListItem>
        <ListItemIcon>
          <Grid4x4Icon color={showGrid ? "info" : ""} />
        </ListItemIcon>
        <ListItemText id="switch-list-label-grid" primary="Grid" />
        <Switch
          edge="end"
          onChange={handleGrid}
          checked={showGrid}
          inputProps={{
            "aria-labelledby": "switch-list-label-grid"
          }}
        />
      </ListItem>
      <ListItem>
        <ListItemIcon>
          <FormatColorFillIcon color="info" />
        </ListItemIcon>
        <ListItemText id="switch-list-label-grid" primary="Colors" />
        <IconButton onClick={handleColorPicker}>
          <ColorLensIcon />
        </IconButton>
      </ListItem>
      {values.map((value) => (
        <ListItem key={value.title}>
          <ListItemIcon>
            <InfoIcon />
          </ListItemIcon>
          <ListItemText id={`${value.title}-label`} primary={value.title} />
          <Typography variant="subtitle1">{value.value}</Typography>
        </ListItem>
      ))}
    </DialogContent>
    <DialogActions sx={{ display: "flex", flexDirection: "row", justifyContent: "space-around" }}>
      <Links />
      <IconButton aria-label="reset" onClick={handleOk}>
        <CancelOutlinedIcon style={{ fontSize: "2rem" }} />
      </IconButton>
    </DialogActions>
  </Dialog>
);

export const Links = () => (
  <>
    <IconButton onClick={() => window.open("https://pi.math.cornell.edu/~lipa/mec/lesson6.html")}>
      <HelpIcon style={{ fontSize: "2rem" }} color="info" />
    </IconButton>

    <IconButton onClick={() => window.open("https://github.com/Dereje1/game-of-life-v2")}>
      <GitHubIcon style={{ fontSize: "2rem" }} color="success" />
    </IconButton>
  </>
);

export default SettingsDialog;

SettingsDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  showGrid: PropTypes.bool.isRequired,
  values: PropTypes.array.isRequired,
  patternName: PropTypes.string.isRequired,
  handleOk: PropTypes.func.isRequired,
  handleGrid: PropTypes.func.isRequired,
  refreshPattern: PropTypes.func.isRequired,
  handleColorPicker: PropTypes.func.isRequired,
  showColorPicker: PropTypes.bool.isRequired
};
