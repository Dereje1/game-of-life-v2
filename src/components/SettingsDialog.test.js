import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import SettingsDialog, { Links } from "./SettingsDialog";

let props;
let windowSpy;
let open = jest.fn();
beforeEach(() => {
  props = {
    open: true,
    values: [
      { title: "Info 1", value: "Info 1 value" },
      { title: "Info 2", value: "Info 2 value" }
    ],
    handleClose: jest.fn(),
    showGrid: true,
    handleGrid: jest.fn(),
    activePattern: "random",
    refreshPattern: jest.fn(),
    handleColorPicker: jest.fn(),
    handlePatternDialog: jest.fn(),
    isColorPickerVisible: false,
    disableColorRestore: true,
    restoreColors: jest.fn()
  };
  windowSpy = jest.spyOn(global, "window", "get");
  windowSpy.mockImplementation(() => ({ open }));
});

afterEach(() => {
  props = null;
  jest.clearAllMocks();
});

test("will render the settings dialog", () => {
  const wrapper = shallow(<SettingsDialog {...props} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test("will render the settings dialog for no pattern or grid", () => {
  const updatedProps = {
    ...props,
    activePattern: "none",
    showGrid: false
  };
  const wrapper = shallow(<SettingsDialog {...updatedProps} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test("will render links", () => {
  const wrapper = shallow(<Links />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

test("will render link to open the information page", () => {
  const wrapper = shallow(<Links />);
  const info = wrapper.find("ForwardRef(IconButton)").at(0);
  info.props().onClick();
  expect(open).toHaveBeenCalledWith("https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life");
});

test("will render link to open the github source page", () => {
  const wrapper = shallow(<Links />);
  const github = wrapper.find("ForwardRef(IconButton)").at(1);
  github.props().onClick();
  expect(open).toHaveBeenCalledWith("https://github.com/Dereje1/game-of-life-v2");
});
