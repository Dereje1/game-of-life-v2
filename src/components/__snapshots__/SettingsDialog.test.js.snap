// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will render links 1`] = `
<Fragment>
  <ForwardRef(IconButton)
    onClick={[Function]}
  >
    <Memo(ForwardRef(HelpIcon))
      color="info"
      style={
        Object {
          "fontSize": "2rem",
        }
      }
    />
  </ForwardRef(IconButton)>
  <ForwardRef(IconButton)
    onClick={[Function]}
  >
    <Memo(ForwardRef(GitHubIcon))
      color="success"
      style={
        Object {
          "fontSize": "2rem",
        }
      }
    />
  </ForwardRef(IconButton)>
</Fragment>
`;

exports[`will render the settings dialog 1`] = `
<ForwardRef(Dialog)
  maxWidth="xs"
  onClose={[MockFunction]}
  open={true}
  sx={
    Object {
      "& .MuiDialog-paper": Object {
        "maxHeight": 530,
        "width": "80%",
      },
    }
  }
>
  <ForwardRef(DialogContent)
    dividers={true}
  >
    <ForwardRef(ListItem)>
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(PatternIcon))
          color="info"
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="pattern-list-label"
        primary="Pattern"
        secondary="Random"
      />
      <ForwardRef(IconButton)
        disabled={false}
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(RefreshIcon))
          color="success"
        />
      </ForwardRef(IconButton)>
      <ForwardRef(IconButton)
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(ManageAccountsIcon))
          color="error"
        />
      </ForwardRef(IconButton)>
    </ForwardRef(ListItem)>
    <ForwardRef(ListItem)>
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(Grid4x4Icon))
          color="info"
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="switch-list-label-grid"
        primary="Grid"
      />
      <ForwardRef(Switch)
        checked={true}
        edge="end"
        inputProps={
          Object {
            "aria-labelledby": "switch-list-label-grid",
          }
        }
        onChange={[MockFunction]}
      />
    </ForwardRef(ListItem)>
    <ForwardRef(ListItem)>
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(FormatColorFillIcon))
          color="info"
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="color-list-label"
        primary="Colors"
      />
      <ForwardRef(IconButton)
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(ColorLensIcon))
          color="success"
        />
      </ForwardRef(IconButton)>
    </ForwardRef(ListItem)>
    <ForwardRef(ListItem)
      key="Info 1"
    >
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(InfoIcon))
          color="warning"
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="Info 1-label"
        primary="Info 1"
      />
      <ForwardRef(Typography)
        variant="subtitle1"
      >
        Info 1 value
      </ForwardRef(Typography)>
    </ForwardRef(ListItem)>
    <ForwardRef(ListItem)
      key="Info 2"
    >
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(InfoIcon))
          color="warning"
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="Info 2-label"
        primary="Info 2"
      />
      <ForwardRef(Typography)
        variant="subtitle1"
      >
        Info 2 value
      </ForwardRef(Typography)>
    </ForwardRef(ListItem)>
  </ForwardRef(DialogContent)>
  <ForwardRef(DialogActions)
    sx={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <Links />
    <ForwardRef(IconButton)
      aria-label="reset"
      onClick={[MockFunction]}
    >
      <Memo(ForwardRef(CancelOutlinedIcon))
        style={
          Object {
            "fontSize": "2rem",
          }
        }
      />
    </ForwardRef(IconButton)>
  </ForwardRef(DialogActions)>
</ForwardRef(Dialog)>
`;

exports[`will render the settings dialog for no pattern or grid 1`] = `
<ForwardRef(Dialog)
  maxWidth="xs"
  onClose={[MockFunction]}
  open={true}
  sx={
    Object {
      "& .MuiDialog-paper": Object {
        "maxHeight": 530,
        "width": "80%",
      },
    }
  }
>
  <ForwardRef(DialogContent)
    dividers={true}
  >
    <ForwardRef(ListItem)>
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(PatternIcon))
          color=""
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="pattern-list-label"
        primary="Pattern"
        secondary="None"
      />
      <ForwardRef(IconButton)
        disabled={true}
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(RefreshIcon))
          color=""
        />
      </ForwardRef(IconButton)>
      <ForwardRef(IconButton)
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(ManageAccountsIcon))
          color="error"
        />
      </ForwardRef(IconButton)>
    </ForwardRef(ListItem)>
    <ForwardRef(ListItem)>
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(Grid4x4Icon))
          color=""
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="switch-list-label-grid"
        primary="Grid"
      />
      <ForwardRef(Switch)
        checked={false}
        edge="end"
        inputProps={
          Object {
            "aria-labelledby": "switch-list-label-grid",
          }
        }
        onChange={[MockFunction]}
      />
    </ForwardRef(ListItem)>
    <ForwardRef(ListItem)>
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(FormatColorFillIcon))
          color="info"
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="color-list-label"
        primary="Colors"
      />
      <ForwardRef(IconButton)
        onClick={[MockFunction]}
      >
        <Memo(ForwardRef(ColorLensIcon))
          color="success"
        />
      </ForwardRef(IconButton)>
    </ForwardRef(ListItem)>
    <ForwardRef(ListItem)
      key="Info 1"
    >
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(InfoIcon))
          color="warning"
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="Info 1-label"
        primary="Info 1"
      />
      <ForwardRef(Typography)
        variant="subtitle1"
      >
        Info 1 value
      </ForwardRef(Typography)>
    </ForwardRef(ListItem)>
    <ForwardRef(ListItem)
      key="Info 2"
    >
      <ForwardRef(ListItemIcon)>
        <Memo(ForwardRef(InfoIcon))
          color="warning"
        />
      </ForwardRef(ListItemIcon)>
      <ForwardRef(ListItemText)
        id="Info 2-label"
        primary="Info 2"
      />
      <ForwardRef(Typography)
        variant="subtitle1"
      >
        Info 2 value
      </ForwardRef(Typography)>
    </ForwardRef(ListItem)>
  </ForwardRef(DialogContent)>
  <ForwardRef(DialogActions)
    sx={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <Links />
    <ForwardRef(IconButton)
      aria-label="reset"
      onClick={[MockFunction]}
    >
      <Memo(ForwardRef(CancelOutlinedIcon))
        style={
          Object {
            "fontSize": "2rem",
          }
        }
      />
    </ForwardRef(IconButton)>
  </ForwardRef(DialogActions)>
</ForwardRef(Dialog)>
`;
